File: backend/app/schemas/kline.py
"""K 线查询 schemas。"""
from __future__ import annotations

from pydantic import BaseModel


class KlineCandle(BaseModel):
    time: str
    open: float
    high: float
    low: float
    close: float
    change_pct: float | None = None
    amount: float | None = None


class KlineMaPoint(BaseModel):
    time: str
    value: float


class KlineVolumePoint(BaseModel):
    time: str
    value: float
    color: str


class KlineMacdBar(BaseModel):
    time: str
    value: float
    color: str


class KlineMacdLine(BaseModel):
    time: str
    value: float


class KlineResponse(BaseModel):
    code: str
    name: str = ""
    adjustment: str = "none"
    candles: list[KlineCandle]
    volume: list[KlineVolumePoint]
    macd: list[KlineMacdBar] = []
    dif: list[KlineMacdLine] = []
    dea: list[KlineMacdLine] = []
    ma5: list[KlineMaPoint]
    ma10: list[KlineMaPoint]
    ma20: list[KlineMaPoint]
    ma60: list[KlineMaPoint]

File: Web/src/api/kline.ts
import { api } from "./client";

export interface KlineCandle {
  time: string;
  open: number;
  high: number;
  low: number;
  close: number;
  change_pct?: number | null;
  amount?: number | null;
}

export interface KlineMaPoint {
  time: string;
  value: number;
}

export interface KlineVolumePoint {
  time: string;
  value: number;
  color: string;
}

export interface KlineMacdBar {
  time: string;
  value: number;
  color: string;
}

export interface KlineMacdLine {
  time: string;
  value: number;
}

export interface KlineResponse {
  code: string;
  name: string;
  adjustment: "qfq" | "none";
  candles: KlineCandle[];
  volume: KlineVolumePoint[];
  macd?: KlineMacdBar[];
  dif?: KlineMacdLine[];
  dea?: KlineMacdLine[];
  ma5: KlineMaPoint[];
  ma10: KlineMaPoint[];
  ma20: KlineMaPoint[];
  ma60: KlineMaPoint[];
}

export async function getKline(
  code: string,
  opts: {
    lastN?: number;
    endDate?: string;
    minDate?: string;
    centerDate?: string;
    maxDate?: string;
    adjust?: "qfq" | "none";
  } = {}
): Promise<KlineResponse> {
  const { data } = await api.get<KlineResponse>(`/kline/${code}`, {
    params: {
      last_n: opts.lastN ?? 300,
      adjust: opts.adjust ?? "qfq",
      ...(opts.endDate ? { end_date: opts.endDate } : {}),
      ...(opts.minDate ? { min_date: opts.minDate } : {}),
      ...(opts.centerDate ? { center_date: opts.centerDate } : {}),
      ...(opts.maxDate ? { max_date: opts.maxDate } : {}),
    },
  });
  return data;
}
